modules/apps/content-dashboard/content-dashboard-journal-impl/src/main/java/com/liferay/content/dashboard/journal/internal/item/action/ViewJournalArticleContentDashboardItemAction.java

```diff
@@ -89,49 +108,146 @@ public String getURL(Locale locale) {
 		return _getViewURL(locale, themeDisplay);
 	}
 
-	private String _getViewURL(Locale locale, ThemeDisplay themeDisplay) {
-		try {
-			ThemeDisplay clonedThemeDisplay =
-				(ThemeDisplay)themeDisplay.clone();
+	private LayoutDisplayPageObjectProvider<JournalArticle>
+		_getLayoutDisplayPageObjectProvider(JournalArticle journalArticle) {
 
-			clonedThemeDisplay.setScopeGroupId(_journalArticle.getGroupId());
+		LayoutDisplayPageProvider<?> layoutDisplayPageProvider =
+			_layoutDisplayPageProviderTracker.
+				getLayoutDisplayPageProviderByClassName(
+					JournalArticle.class.getName());
 
-			return Optional.ofNullable(
-				_assetDisplayPageFriendlyURLProvider.getFriendlyURL(
+		if (layoutDisplayPageProvider == null) {
+			return null;
+		}
+
+		return (LayoutDisplayPageObjectProvider<JournalArticle>)
+			layoutDisplayPageProvider.getLayoutDisplayPageObjectProvider(
+				new InfoItemReference(
 					JournalArticle.class.getName(),
-					_journalArticle.getResourcePrimKey(), locale,
-					clonedThemeDisplay)
-			).map(
-				url -> {
-					String backURL = ParamUtil.getString(
-						_httpServletRequest, "backURL");
-
-					if (Validator.isNotNull(backURL)) {
-						return HttpComponentsUtil.setParameter(
-							url, "p_l_back_url", backURL);
-					}
+					journalArticle.getResourcePrimKey()));
+	}
 
-					return HttpComponentsUtil.setParameter(
-						url, "p_l_back_url", themeDisplay.getURLCurrent());
-				}
-			).orElse(
-				StringPool.BLANK
-			);
-		}
-		catch (CloneNotSupportedException | PortalException exception) {
-			_log.error(exception);
+	private Optional<Layout> _getLayoutOptional(
+		LayoutDisplayPageObjectProvider<?> layoutDisplayPageObjectProvider) {
+
+		return Optional.ofNullable(
+			layoutDisplayPageObjectProvider
+		).filter(
+			currentLayoutDisplayPageObjectProvider ->
+				currentLayoutDisplayPageObjectProvider.getDisplayObject() !=
+					null
+		).map(
+			currentLayoutDisplayPageObjectProvider ->
+				AssetDisplayPageUtil.getAssetDisplayPageLayoutPageTemplateEntry(
+					layoutDisplayPageObjectProvider.getGroupId(),
+					layoutDisplayPageObjectProvider.getClassNameId(),
+					layoutDisplayPageObjectProvider.getClassPK(),
+					layoutDisplayPageObjectProvider.getClassTypeId())
+		).map(
+			layoutPageTemplateEntry -> _layoutLocalService.fetchLayout(
+				layoutPageTemplateEntry.getPlid())
+		);
+	}
 
-			return StringPool.BLANK;
-		}
+	private String _getLocalizedURL(
+		List<LayoutSEOLink> localizedLayoutSEOLinks, Locale locale) {
+
+		List<LayoutSEOLink> result = new ArrayList<>();
+
+		ListUtil.filter(
+			localizedLayoutSEOLinks, result,
+			seoLink -> {
+				String stringLocale = locale.toString();
+
```

Comment:
You can compare Locales directly :)

Reply:
Once we have the seoLink, the HrefLang is a String, not a Locale. To compare them I have to write all that process

Reply:
Take a look to Language class, I think probably there is a method there you can use